### PR TITLE
Strings / it-academy.by / JC1-110-22 / Alena Trafimovich / Done tasks 1-18

### PR DESCRIPTION
**Pull request by AlenaTrafimovich**
## Pull request state

Actual status for PR:
- [x] implemented
- [x] reviewed
- [x] fixed
- [x] approved

### CODE REVIEW

File: src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java [hunk 1]
```diff
@@ -0,0 +1,244 @@
+package com.rakovets.course.java.core.practice.strings;
+
+import java.util.Arrays;
+
+public class StringUtil {
+
+    //Task 1
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java:7]: "Violation: Useless comments
Solution: Remove comments"

File: src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+        } else if (str1 == null) {
+            return str2;
+        } else if (str2 == null) {
+            return str1;
+        } else {
+            return str1.concat(str2);
+        }
+    }
+
+    //Task 2
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java:20]: "Violation: Useless comments
Solution: Remove comments"

File: src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        }
+    }
+
+    //Task 4
+    public String getTrimSpaceToLowerCase(String str) {
+        return str == null ? null : str.trim().toLowerCase();
+    }
+
+    //Task 5
+    // excluding second symbol
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java:40]: "Violation: Useless comments
Solution: Remove comments"

File: src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+            return null;
+        } else if (oldValue == null) {
+            return str;
+        } else if (newValue == null) {
+            newValue = "";
+        }
+        return str.replace(oldValue, newValue);
+    }
+
+    //Task 7
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java:59]: "Violation: Useless comments
Solution: Remove comments"

File: src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    public boolean isStartAndEndWithWord(String text, String word) {
+        return text != null && word != null && text.startsWith(word) && text.endsWith(word);
+    }
+
+    //Task 8
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java:64]: "Violation: Useless comments
Solution: Remove comments"

File: src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+            for (char vowel : vowels) {
+                if (c == vowel) {
+                    amountVowels++;
+                }
+            }
+        }
+        return amountVowels;
+    }
+
+    //Task 9
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java:82]: "Violation: Useless comments
Solution: Remove comments"

File: src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+            for (char punctuationMark : punctuationMarks) {
+                if (c == punctuationMark) {
+                    amountPunctuationMarks++;
+                }
+            }
+        }
+        return amountPunctuationMarks;
+    }
+
+    //Task 10
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java:100]: "Violation: Useless comments
Solution: Remove comments"

File: src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+        if (str == null) {
+            return false;
+        }
+        String modified = str.toLowerCase().replaceAll("\\W", "");
+        StringBuilder stringBuilder = new StringBuilder(modified);
+        String inverted = stringBuilder.reverse().toString();
+        return isEqual(modified, inverted);
+    }
+
+    //Task 11
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java:111]: "Violation: Useless comments
Solution: Remove comments"

File: src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+        if (str == null) {
+            return null;
+        }
+        int length = 0;
+        if (str.length() % numberSymbols == 0) {
+            length += str.length() / numberSymbols;
+        } else {
+            length += str.length() / numberSymbols + 1;
+        }
+        String[] array = new String[length];
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java:122]: useless code

File: src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+            if (i == length - 1) {
+                array[i] = str.substring(i * numberSymbols);
+            } else {
+                array[i] = str.substring(i * numberSymbols, i * numberSymbols + numberSymbols);
+            }
+        }
+        return array;
+    }
+
+    //Task 12
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java:133]: "Violation: Useless comments
Solution: Remove comments"

File: src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    public int countNumberWords(String str) {
+        return str == null ? 0 : str.trim().replaceAll("[\\pP\\s]", " ").split(" +").length;
+    }
+
+    //Task 13
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java:138]: "Violation: Useless comments
Solution: Remove comments"

File: src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+    public String getFirstCharacters(String str) {
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java:139]: "Violation: Wrong method name
Solution: Rename method"

File: src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+            for (char number : numbers) {
+                if (symbol == number) {
+                    stringBuilder.append(symbol);
+                }
+            }
+        }
+        return stringBuilder.toString();
+    }
+
+    //Task 15
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java:168]: "Violation: Useless comments
Solution: Remove comments"

File: src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java [hunk 14]
```diff
@@ -0,0 +1,487 @@
...
+    void testStartAndEndWithWordNullWord() {
+        verifyStartAndEndWithWord("hello", null, false);
+    }
+
+    void verifyStartAndEndWithWord(String str1, String word, Boolean expected) {
+        Boolean actual = stringUtil.isStartAndEndWithWord(str1, word);
+        Assertions.assertEquals(expected, actual);
+    }
+
+    //Tests for Task8
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java:226]: "Violation: Useless comments
Solution: Remove comments"

File: src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java [hunk 15]
```diff
@@ -0,0 +1,487 @@
...
+    void testCountVowelsNullString() {
+        verifyCountVowels(null, 0);
+    }
+
+    void verifyCountVowels(String str, int expected) {
+        int actual = stringUtil.countVowels(str);
+        Assertions.assertEquals(expected, actual);
+    }
+
+    //Tests for Task9
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java:248]: "Violation: Useless comments
Solution: Remove comments"

File: src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java [hunk 16]
```diff
@@ -0,0 +1,487 @@
...
+    void testPunctuationMarksNullString() {
+        verifyPunctuationMarks(null, 0);
+    }
+
+    void verifyPunctuationMarks(String str1, int expected) {
+        int actual = stringUtil.countPunctuationMarks(str1);
+        Assertions.assertEquals(expected, actual);
+    }
+
+    //Tests for Task10
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java:265]: "Violation: Useless comments
Solution: Remove comments"

File: src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java [hunk 17]
```diff
@@ -0,0 +1,487 @@
...
+    void testSplitToArrayStringNull() {
+        verifySplitToArray(null, 3, null);
+    }
+
+    void verifySplitToArray(String str1, int numberSymbolsInEveryNewArray, String[] expected) {
+        String[] actual = stringUtil.splitToArray(str1, numberSymbolsInEveryNewArray);
+        Assertions.assertArrayEquals(expected, actual);
+    }
+
+    //Tests for Task12
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java:324]: "Violation: Useless comments
Solution: Remove comments"

File: src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java [hunk 18]
```diff
@@ -0,0 +1,487 @@
...
+    void testCountNumberWordsStringNull() {
+        verifyCountNumberWords(null, 0);
+    }
+
+    void verifyCountNumberWords(String str1, int expected) {
+        int actual = stringUtil.countNumberWords(str1);
+        Assertions.assertEquals(expected, actual);
+    }
+
+    //Tests for Task13
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java:351]: "Violation: Useless comments
Solution: Remove comments"

File: src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java [hunk 19]
```diff
@@ -0,0 +1,487 @@
...
+    void testGetFirstCharactersNullString() {
+        verifyGetFirstCharacters(null, "");
+    }
+
+    void verifyGetFirstCharacters(String str1, String expected) {
+        String actual = stringUtil.getFirstCharacters(str1);
+        Assertions.assertEquals(expected, actual);
+    }
+
+    //Tests for Task14
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java:373]: "Violation: Useless comments
Solution: Remove comments"

File: src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java [hunk 20]
```diff
@@ -0,0 +1,487 @@
...
+    void testGetNumbersNullString() {
+        verifyGetNumbers(null, "");
+    }
+
+    void verifyGetNumbers(String str, String expected) {
+        String actual = stringUtil.getNumbers(str);
+        Assertions.assertEquals(expected, actual);
+    }
+
+    //Tests for Task15
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java:390]: "Violation: Useless comments
Solution: Remove comments"

File: src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java [hunk 21]
```diff
@@ -0,0 +1,487 @@
...
+    void testIsArraySimilarDigitsIdentical() {
+        verifyIsArraySimilar(new String[]{"56", "28"}, new String[]{"28", "56"}, true);
+    }
+
+    void verifyIsArraySimilar(String[] array1, String[] array2, boolean expected) {
+        boolean actual = stringUtil.isArraySimilar(array1, array2);
+        Assertions.assertEquals(expected, actual);
+    }
+
+    //Tests for Task17
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java:449]: "Violation: Useless comments
Solution: Remove comments"

File: src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java [hunk 22]
```diff
@@ -0,0 +1,487 @@
...
+    void testGetSpeedOfAddition2() {
+        System.out.println(stringUtil.getSpeedOfAddition("10", 15));
+    }
+
+    @Test
+    void testGetSpeedOfAddition3() {
+        System.out.println(stringUtil.getSpeedOfAddition("10", 20));
+    }
+
+    //Tests for Task18
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/strings/StringUtilTest.java:466]: "Violation: Useless comments
Solution: Remove comments"

**Review by rakovets — CHANGES_REQUESTED** [review]
Need to fix

File: src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java (hunk 9 above)
Comment (AlenaTrafimovich) [reply src/main/java/com/rakovets/course/java/core/practice/strings/StringUtil.java:122]: Could you please explain why is useless, cause without this condition if I have even amount of symbols in my String I get StringIndexOutOfBoundsException Error.

**Comment by AlenaTrafimovich** [comment]
fixed review comments

**Review by rakovets — APPROVED** [review]

**Pull request closed without merging**